### ISSUE TITLE
Only first line of figure description extracted if distance between lines deemed too large

### ISSUE THREAD

**Issue by de-code** [kermitt2/grobid#683]
After having re-trained the `fulltext` and `figure` GROBID models, I noticed the extracted figure description / caption may sometimes only contain the first line despite the (DL) model correctly tagging the full description.

`482984v1` (DOI: [10.1101/482984](https://doi.org/10.1101/482984)) (from the bioRxiv 10k validation set, not to be added to training)

<details>
<summary>PDF</summary>

![image](https://user-images.githubusercontent.com/1016473/102130097-4cae7080-3e48-11eb-8af8-53d4d5f36fda.png)
</details>

<details>
<summary>target xml</summary>

```xml
<fig xmlns:mml="http://www.w3.org/1998/Math/MathML" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" id="fig1" position="float" fig-type="figure">
<label>Figure 1.</label>
<caption><title>Regression analysis of mesophyll conductance (<italic>g</italic><sub>m</sub>) with photosynthetic-N use efficiency (PNUE) in four species seedling leaves. The determination coefficient (<italic>R</italic><sup>2</sup>) and <italic>P</italic>-value are also shown. The lines fitted separately for four species were significantly different (<italic>P</italic> < 0.05) according to the result of a one-way ANCOVA with PNUE as a dependent variable, tree species as fixed factors, and <italic>g</italic><sub>m</sub> as a covariate.</title></caption>
<graphic xlink:href="482984_fig1.tif"/>
</fig>

<fig xmlns:mml="http://www.w3.org/1998/Math/MathML" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" id="fig2" position="float" fig-type="figure">
<label>Figure 2.</label>
<caption><title>Regression analysis of the fraction of leaf N allocated to (a) Rubisco (<italic>P</italic><sub>R</sub>), (b) bioenergetics (<italic>P</italic><sub>B</sub>), (c) light-harvesting components (<italic>P</italic><sub>L</sub>), and (d) the photosynthetic apparatus (<italic>P</italic><sub>P</sub>) with photosynthetic-N use efficiency (PNUE) in four species seedling leaves. The determination coefficient (<italic>R</italic><sup>2</sup>) and <italic>P</italic>-value are also shown. Only one line is fitted for four species, because there was no significant difference (<italic>P</italic> > 0.05) according to the result of a one-way ANCOVA with PNUE as a dependent variable, tree species as fixed factors, and <italic>P</italic><sub>R</sub><italic>, P</italic><sub>B</sub><italic>, P</italic><sub>L,</sub> or <italic>P</italic><sub>P</sub> as a covariate.</title></caption>
<graphic xlink:href="482984_fig2.tif"/>
</fig>

<fig xmlns:mml="http://www.w3.org/1998/Math/MathML" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" id="fig3" position="float" fig-type="figure">
<label>Figure 3.</label>
<caption><title>Regression analysis of the fraction of leaf N allocated to the cell wall (<italic>P</italic><sub>CW</sub>) with leaf N allocated to Rubisco (<italic>P</italic><sub>R</sub>) in four species seedling leaves. The determination coefficient (<italic>R</italic><sup>2</sup>) and <italic>P</italic>-value are also shown. The shaded zone represents the distribution area of <italic>P</italic><sub>CW</sub> and <italic>P</italic><sub>R</sub> when a trade-off exists, for more information see Harrison <italic>et al</italic>. [<xref rid="c18" ref-type="bibr">18</xref>]. The lines fitted separately for four species were significantly different (<italic>P</italic> < 0.05) according to the result of a one-way ANCOVA with <italic>P</italic><sub>R</sub> as a dependent variable, tree species as fixed factors, and <italic>P</italic><sub>CW</sub> as a covariate.</title></caption>
<graphic xlink:href="482984_fig3.tif"/>
</fig>
```
</details>

<details>
<summary>GROBID TEI XML</summary>

```xml
<figure xmlns="http://www.tei-c.org/ns/1.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xlink="http://www.w3.org/1999/xlink" xml:id="fig_1">
  <head>Figure 1.</head>
  <label>Figure 1.</label>
  <figDesc>Figure 1. Regression analysis of mesophyll conductance (g m ) with photosynthetic-N use efficiency (PNUE) in</figDesc>
</figure>

<figure xmlns="http://www.tei-c.org/ns/1.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xlink="http://www.w3.org/1999/xlink" xml:id="fig_2">
  <head>Figure 2.</head>
  <label>Figure 2.</label>
  <figDesc>Figure 2. Regression analysis of the fraction of leaf N allocated to (a) Rubisco (P R ), (b) bioenergetics (P B ), (c)</figDesc>
</figure>

<figure xmlns="http://www.tei-c.org/ns/1.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xlink="http://www.w3.org/1999/xlink" xml:id="fig_3">
  <head>Figure 3.</head>
  <label>Figure 3.</label>
  <figDesc>Figure 3. Regression analysis of the fraction of leaf N allocated to the cell wall (P CW ) with leaf N allocated to</figDesc>
</figure>
```
</details>

The DL model actually seem to receive and label more text as the description, e.g.

<details>
<summary>ScienceBeam DeLFT tagging output</summary>

```xml
<xml>
  <p>
    <label>Figure 1 .</label>
    <figDesc>Regression analysis of mesophyll conductance ( g m ) with photosynthetic - N use efficiency ( PNUE ) in four species seedling leaves . The determination coefficient ( R 2 ) and P - value are also shown . The lines fitted separately for four species were significantly different ( P &lt; 0 . 05 ) according to the result of a one - way ANCOVA with PNUE as a dependent variable , tree species as fixed factors , and g m as a covariate .</figDesc>
  </p>
</xml>
```
</details>

---

This seem to be down to `org.grobid.core.document.Document.getFigureLayoutTokens` (https://github.com/kermitt2/grobid/blob/0.6.1/grobid-core/src/main/java/org/grobid/core/document/Document.java#L1489).

In the case of the first extracted figure (see above), the following expression
`BoundingBox.fromPointAndDimensions(b.getPageNumber(), b.getX(), b.getY(), b.getWidth(), b.getHeight()).distanceTo(prevBlock)` equates to `21.698000`.

Perhaps it would be good if there was an option to disable that logic? (in cases where there is enough training data)

**Comment by kermitt2** [comment]
Hi Daniel !

Yes this rule was introduced as a "quick" security to avoid frequent errors and it works well in general for standard publication. At the time we were not really considering review papers with extra-large line spacing, which are a low proportion of papers. 

This is easy to fix and get it a bit more general, we simply need to use the average/median line space separation of the document instead of the fixed value `15` (it will be much faster actually to add this than such an obscure option that will need to be documented/maintained, which does not make sense imho). It was anyway supposed to be a temporary hack because of the lack of training data and reliability of the full text model. The more medium term objective would be to have this median line spacing value as a feature in the full text model to prevent this situation earlier.

**Comment by de-code** [comment]
Hi Patrice, thank you for your response.

> This is easy to fix and get it a bit more general, we simply need to use the average/median line space separation of the document

It seems like it should be easy. I am slightly hesitant putting all my faith into it being simple given the issues with the character spacing. (e.g. with a mixture of font sizes)

Adding it as a feature to the model seems to be a good way.

Perhaps what I would propose is to:

- first add a feature flag to disable that safety net altogether.
- Then we could do more tests, whether we would get much value out of adding that additional feature.

I used feature flags in other circumstances, as they are quite light-weight and allow a step-wise transition to a new feature / behaviour, or just a trial. Feature flags do not have a long-term commitment.

In general it would probably be good if most models just received the same data. (e.g. the same feature could be useful to the `fulltext` model).

**Comment by de-code** [comment]
BTW The similar area of code (`assignGraphicObjectsToFigures`: https://github.com/kermitt2/grobid/blob/0.6.1/grobid-core/src/main/java/org/grobid/core/document/Document.java#L1153-L1159) seems to also be responsible for, in my case, include the label / header in the description. I haven't yet reproduced that with the default models, I guess there is some code handling it. In my model the label and header are the same (e.g. `Figure 1`).

There doesn't seem to be the same logic for tables even though they seem to be very similar.

**Comment by de-code** [comment]
Just to get an idea what difference this can make, an evaluation over the bioRxiv 10k validation dataset (first 200 samples):

![image](https://user-images.githubusercontent.com/1016473/104202071-5d8bdb00-5422-11eb-82a7-2eaf7f110a2a.png)

The last two bars are using the same models. The difference being that for the last bar, the rules relating to updating the figure caption have been disabled (via https://github.com/elifesciences/grobid/pull/32)

**Comment by de-code** [comment]
I guess one counter example, where rules or an additional model input could help:

`475335v1` (DOI: [10.1101/475335](https://doi.org/10.1101/475335)) (bioRxiv 10k train example)

<details>
<summary>PDF</summary>

![image](https://user-images.githubusercontent.com/1016473/104206055-04727600-5427-11eb-95e1-e4540fe228c7.png)
</details>

<details>
<summary>bioRxiv JATS XML</summary>

```xml
<sec id="s3c">
<title>Epidemic synchrony and annual phase coherence</title>
<p>
We explored correlations between dengue time series in different regions. Both epidemic synchrony and phase coherence were higher for closer regions and declined with distance (
<xref rid="fig3" ref-type="fig">Fig 3</xref>
). For the Urban-2 (n = 161) spatial level, epidemic synchrony reached the average countrywide correlation at approximately 1,260 kilometres (
<xref rid="fig3" ref-type="fig">Fig 3A</xref>
). This synchrony length represents a substantial part of Brazil’s dimensions as the country extends 4,395 kilometres north to south and 4,319 kilometres west to east. The coherence length had a higher value of 1,590 kilometres (
<xref rid="fig3" ref-type="fig">Fig 3B</xref>
), suggesting that agreement in dengue seasonality spreads further than correlations of epidemic curves.
</p>
<fig id="fig3" position="float" fig-type="figure">
<label>Fig 3.</label>
<caption>
<title>
Epidemic synchrony and annual phase coherence between Brazilian Urban-2 regions.
</title>
<p>
Epidemic synchrony (A) and annual phase coherence (B) summarised using nonparametric spline covariance function. Solid blue line describes the mean pairwise correlation from the data and the dotted lines represent the 95% envelope for bootstrapped correlations of case and annual phase angle time series, respectively. Red line indicates global countrywide correlation.
</p>
</caption>
<graphic xlink:href="475335_fig3.tif"/>
</fig>
<p>
We also looked at epidemic synchrony and annual phase coherence at other spatial levels (S4 and S5 Figs, respectively) and found that both synchrony and coherence lengths tend to decrease for smaller spatial resolutions and stabilise at 1,240 km and 1,500 km.
</p>
</sec>
```
</details>

<details>
<summary>GROBID TEI XML (with re-trained models)</summary>

```xml
<figure xmlns="http://www.tei-c.org/ns/1.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xlink="http://www.w3.org/1999/xlink" xml:id="fig_2">
  <head>Fig 3.</head>
  <label>Fig 3.</label>
  <figDesc>Epidemic synchrony and annual phase coherence between Brazilian Urban-2 regions. Epidemic synchrony (A) and annual phase coherence (B) summarised using nonparametric spline covariance function. Solid blue line describes the mean pairwise correlation from the data and the dotted lines represent the 95% envelope for bootstrapped correlations of case and annual phase angle time series, respectively. Red line indicates global countrywide correlation. We also looked at epidemic synchrony and annual phase coherence at other spatial levels (S4 and S5 Figs, respectively) and found that both synchrony and coherence lengths tend to decrease for smaller spatial resolutions and stabilise at 1,240 km and 1,500 km.</figDesc>
</figure>
```
</details>

**Comment by lfoppiano** [comment]
As a complementary information, I removed the logic and ran the end 2 end evaluation. I saw no changes in both evaluations. I've tested on a slightly different data set, though.

**Comment by kermitt2** [comment]
> As a complementary information, I removed the logic and ran the end 2 end evaluation. I saw no changes in both evaluations. I've tested on a slightly different data set, though.

The end-2-end evaluation only covers the figure title extraction, not the figure caption, so no difference is to be expected. 

This "security" was introduced from the feedback of ResearchGate on processing very large volume of figure extractions (several millions), this remains not so frequent, but having incorrect extra-long figure caption has a very bad impact on users. 

The figure label added in the figure description was also a decision from ResearchGate feedback, making the extraction in general more readable (although I was not found of that personally).

**Comment by lfoppiano** [comment]
> The end-2-end evaluation only covers the figure title extraction, not the figure caption, so no difference is to be expected.

oh, I missed this completely. What is the difference between figure title and figure caption?

**Comment by kermitt2** [comment]
> oh, I missed this completely. What is the difference between figure title and figure caption?

There is often a title to the figure (often in bold) and a distinct description blabla (caption), and usually some sort of label/identifier for the cross-reference in the text. This is the general pattern considered in the annotations for the training data. 

The difference between figure title and figure caption is generally relatively clear, but not always. For instance below... no real title in this case, except `Fig. 1` (is in this case redundant with the label) and some caption, which could easily be interpreted as figure title as well. In the PMC JATS, if I remember correctly, this is not encoded in a consistent manner, so I ignored figure captions in the end-2-end eval.  

```xml
<figure
                xmlns="http://www.tei-c.org/ns/1.0" xml:id="fig_0">
                <head>Fig 1 .</head>
                <label>1</label>
                <figDesc>
                    <div>
                        <p>Fig 1. CXR patterns. A) left lower zone consolidation; B) bilateral lower zone interstitial infiltrates; F) scattered alveolar infiltrates in right lung; H) bilateral diffuse consolidation. CXR Severity Scoring (Right + Left): (A) 0+1 = 1; (B) 1+1 = 2; (C) 2+1 = 3; (D) 2+2 = 4; (E) 2+3 = 5; (F) 4+2 = 6; (G) 4+3 = 7; (H) 4 +4 = 8.</p>
                    </div>
                </figDesc>
            </figure>
```

**Comment by de-code** [comment]
For what it's worth, in the bioRxiv dataset there doesn't seem to be a separate title as far as I am aware. I am just evaluating `label` and `caption` / `description` (where I use the full label, e.g. `Fig 1`).

> The figure label added in the figure description was also a decision from ResearchGate feedback, making the extraction in general more readable (although I was not found of that personally).

But since it is a separate field, this seems like a display decision? Perhaps it's to also include the separator, e.g. `.` or `:`? Although that seems to be part of the `head`.

It's not done for tables anyway. (I am not sure why the figures and tables are not just sharing more code).

**Comment by de-code** [comment]
I could look into adding something to the training data...

In general I prefer more flexibility in terms of data being used by the model, closer to the model itself. Or in other words make more raw data available to the model and do the preprocessing then (see #543 and #544). Although that is probably more up for debate and shouldn't be rushed through.

We could for example add a relative line spacing feature. e.g. `2.0` if the line space between the current line and the previous line is twice as big. I would add it as a float. Although the current feature support in DeLFT is geared towards categorical values. So maybe it could be rounded to something like `1.0 or less`, `1.5` and `2.0 or more`.
I am not sure how well that would work for two column layouts or text across pages.

The model could then still come up with his own answer and ignore that parameter.

Any thoughts?
